annotated.go

```diff
@@ -189,13 +288,170 @@ func (rt resultTagsAnnotation) apply(ann *annotated) error {
 	return nil
 }
 
+// build builds and returns a constructor after applying a ResultTags annotation
+func (rt resultTagsAnnotation) build(ann *annotated) (interface{}, error) {
+	paramTypes := ann.currentParamTypes()
+
+	resultTypes, remapResults := rt.results(ann)
+
+	origFn := reflect.ValueOf(ann.Target)
+	newFnType := reflect.FuncOf(paramTypes, resultTypes, false)
+	newFn := reflect.MakeFunc(newFnType, func(args []reflect.Value) []reflect.Value {
+		results := origFn.Call(args)
+		return remapResults(results)
+	})
+	return newFn.Interface(), nil
+}
+
+// results returns the types of the results of the annotated function,
+// and a function that maps the results of the target function,
+// into a result compatible with the annotated function.
+func (rt resultTagsAnnotation) results(ann *annotated) (
+	types []reflect.Type,
+	remap func([]reflect.Value) []reflect.Value,
+) {
+	types, hasError := ann.currentResultTypes()
+
+	if hasError {
+		types = types[:len(types)-1]
+	}
+
+	// No result annotations. Return the original types
+	// and an identity function.
+	if len(rt.tags) == 0 {
+		return types, func(results []reflect.Value) []reflect.Value {
+			return results
+		}
+	}
+
+	// if there's no Out struct among the return types, there was no As annotation applied
+	// just replace original result types with an Out struct and apply tags
+	var (
+		newOut       outStructInfo
+		existingOuts []reflect.Type
+	)
+
+	newOut.Fields = []reflect.StructField{
+		{
+			Name:      "Out",
+			Type:      reflect.TypeOf(Out{}),
+			Anonymous: true,
+		},
+	}
+	newOut.Offsets = []int{}
+
+	for i, t := range types {
+		if isOut(t) {
+			// this must be from an As annotation
+			// apply the tags to the existing type
+			taggedFields := make([]reflect.StructField, t.NumField())
+			taggedFields[0] = reflect.StructField{
+				Name:      "Out",
+				Type:      reflect.TypeOf(Out{}),
+				Anonymous: true,
+			}
+			for j, tag := range rt.tags {
+				if j+1 < t.NumField() {
+					field := t.Field(j + 1)
+					taggedFields[j+1] = reflect.StructField{
+						Name: field.Name,
+						Type: field.Type,
+						Tag:  reflect.StructTag(tag),
+					}
+				}
+			}
+			existingOuts = append(existingOuts, reflect.StructOf(taggedFields))
+			continue
+		}
+		// this must be from the original function
+		// apply the tags
+		field := reflect.StructField{
+			Name: fmt.Sprintf("Field%d", i),
+			Type: t,
+		}
+		if i < len(rt.tags) {
+			field.Tag = reflect.StructTag(rt.tags[i])
+		}
+		newOut.Offsets = append(newOut.Offsets, len(newOut.Fields))
+		newOut.Fields = append(newOut.Fields, field)
+		continue
```

Comment:
Is this continue necessary? It's at the end of the loop.

Reply:
Removed